Develop

drnimbusrain opened the pull request:
* Three new modules:
  - canopy_zpd_mod -- zero plane displacement height and total surface roughness length basd on Massman et al. (2017)
  - canopy_eddyx_mod -- calculates the estimated eddy diffusivity coefficient (Kest(z/hc)) for vertical diffusion of tracers within the canopy based on Makar et al. (2017). Note:  This should be used to scale the first model layer resolved model diffusivity (Kmod(z1)) to get canopy diffusivity (Kcan(z)) such as     Kcan(z) = [Kmod (z1)/Kest(z1/hc)]  * Kest(z/hc)]
  - canopy_phot_mod -- photolysis attenuation, currently Beer's law direct beam attenuation as used in Makar et al. (2017)
* Tables of input parameter and settings in README
* Fixed bugs and other issues
* Added a LU_OPT in namelist for using different land use/veg mapping (currently only VIIRS LU)
* Added user options for LAI, forest fraction, and canopy height thresholds for contiguous forests
* Added simple stem (woody) area index for PAI calculation (PAI = LAI + WAI)
* Added MOST and RSL effects to above canopy wind profile.

zmoon commented on canopy_waf_mod.F90:
```diff
@@ -102,7 +102,7 @@ SUBROUTINE CANOPY_WAF( HCM, ZTOTHC, FAFRACK, FAFRACK0, UBZREF, Z0GHCM, &
                 ( 1.0 + (delta*log((1.0/delta) + 1.0)) )
             term2 = log( LAMDARS * ( ( ( (HREF/HCM) + 1.0 - d_h) )/ zo_h ) )  !denominator
-            waf   = term1 / term2
+            waf   = abs (term1 / term2) !need absolute for flameh < 1 m conditions
```
If this necessarily the case? Doesn't seem to be.

![image](https://user-images.githubusercontent.com/15079414/186259115-8b33182d-c228-4646-9fe7-c32bf2c6126f.png)

<details><summary>Code</summary>

```python
import matplotlib.pyplot as plt
import numpy as np

lrs = 1.25  # coeff; usually this value is used
# lrs = 1.07

h = 20  # canopy height
d_h = 0.8  # zero-plane displacement height (relative)
d = h * d_h  # displacement height
H = 6  # reference height minus h
z0_h = 0.1  # z0/h ranges from 0 to 0.22 in Fig 6
z0 = z0_h * h  # roughness length

# h_flame: flame height rt. h
h_flame = np.linspace(0.01, h, 200)

delta = (1 - d_h) / (h_flame/h)

num = (
    np.log(lrs * (h_flame/h + 1 - d_h) / z0_h)
    - 1
    + delta * np.log(1/delta + 1)
)
den = np.log(lrs * (H/h + 1 - d/h) / z0_h)

waf = num/den
# waf = np.abs(waf)

fig, [ax1, ax2] = plt.subplots(2, 1, sharex=True, figsize=(4, 4))

ax1.plot(h_flame, num)
ax1.set(title=f"Numerator; denominator = {den:.3g}")

ax2.plot(h_flame, waf)
ax2.set(xlabel="$h_\mathrm{flame}$", title="WAF")
ax2.set_xlim(xmin=0)

for ax in fig.get_axes():
    ax.axvline(1, c="0.7")

fig.tight_layout()
```
</details>

The original formula doesn't have abs. Maybe negative WAF value indicates problematic input and should set to 1 or some sort of flag value instead.

drnimbusrain replied:
From my quick checking, it appears that flameh<1 is setting this off, but your right its sloppy.  Needed to further diagnose and fix to avoid negative WAF.

zmoon replied:
I think this is resolved now

zmoon commented on canopy_waf_mod.F90:
```diff
@@ -102,7 +102,7 @@ SUBROUTINE CANOPY_WAF( HCM, ZTOTHC, FAFRACK, FAFRACK0, UBZREF, Z0GHCM, &
             term1 = log( LAMDARS * ( ( (FLAMEH/HCM) +  1.0 - d_h ) / zo_h ) ) - &   !numerator
                 ( 1.0 + (delta*log((1.0/delta) + 1.0)) )
```
```suggestion
                1.0 + (delta*log((1.0/delta) + 1.0))
```

![image](https://user-images.githubusercontent.com/15079414/186260118-ac1eddd0-5d50-4b5d-b298-fde6ee8576ed.png)

(unless it is wrong in the paper)

This could explain why getting negative values.

drnimbusrain replied:
Nice catch!  Yes, that is why.  Thanks Zach.

drnimbusrain commented on canopy_driver.F90:
```diff
@@ -210,13 +211,32 @@ program canopy_driver
+                if (nlon .eq. 1) then      !single lon point -- reset to user value
+                    flameh = flameh_set
+                else                       !calculate flameh
+                    dx     = dlon*111000.0  !convert lon grid to distance (m)
```
OK, we can switch to latitude maybe later, as the input_variables file has longitude always varying.  I agree to move dx up, and take absolute to avoid possible negatives. Better?

zmoon replied:
Yes, it is more robust now. What I was alluding to though is that the value of `dx` in such a lat/lon grid changes with latitude. And we could account for that. But maybe that can wait until we set up to work with 2-d array inputs (e.g. loaded from netCDF files).

drnimbusrain commented on canopy_utils_mod.F90:
```diff
@@ -70,5 +70,23 @@ function CalcPAI(fch, ffrac)
+        real(rk), intent(in)  :: dx           !! Input Grid cell length (m)
+        real(rk)              :: CalcFlameH   !! Calculated Plant area index (PAI)
+
+        CalcFlameH=0.0775*((frp*1000.0)/dx)**0.46  !Byram flameh calculation as function of FRP
```
It is true, and I modify the dlon calc to take absolute value now first,  so that dx is never negative.    However, dx = 0 can be valid if we are running this for a single point, i.e., nlon = 1.  I have this condition covered later, which sets back to the user flameh_set value in the namelist, as if there is no fire/FRP intensity over model grid distance flameh cannot be calculated.
...
...
if (nlon .eq. 1 .or. dx .eq. 0) then      !single lon point -- reset to user value
                    flameh = flameh_set

drnimbusrain commented:
@zmoon there is already an open review pending for this PR, and no way to send reminder/refresh really.  Your initial comments on this PR were spot on, and I know I added more commits, but they should be relatively straightforward.  Do you mind reviewing this sometime later next week or the following?

zmoon commented:
@drnimbusrain I guess the re-request button isn't showing because I still have a pending one like you pointed out. For sure I can review later next week!

zmoon commented:
The namelist param info in readme might work better as tables. I like the idea though.

drnimbusrain commented:
@zmoon Thank you for the README comment.  I also like the use of tables as well.  Let me know.

zmoon commented on canopy_phot_mod.F90:
```diff
@@ -0,0 +1,53 @@
+contains
+
+!:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
+    SUBROUTINE CANOPY_PHOT( HCM, ZK, FAFRACK, LAI, CLU, COSZEN, RJCORR )
```
just making them `elemental` might be sufficient to be able to call with the full profile

drnimbusrain replied:
@zmoon I agree on all accounts, and I have eliminated the need for HCM and ZK and pass in full profiles as the calculation naturally accounts for above and below canopy.  You can see these changes as well as the change in RJCORR variable name, and explicit information on Eq. and page number for the photolysis attenuation calculation in the next [commit](https://github.com/noaa-oar-arl/canopy-app/pull/20/commits/580eca3f68f58426c9ba66dc36a0fa8a66c4bf5b), and can resolve the conversation if satisfied.

However, seems like the canopy_wind calculation will need to be called separately somewhere anyway because of multiple conditions in the submodule that are not only controlled by FAFRACK profile, e.g.,:

       zkhcm = ZK/HCM
        z0g = Z0GHCM*HCM
        hol = HCM/MOL
       ! Nondimensional canopy wind speed term that dominates near the ground:
        if (ZK >= z0g .and. ZK <= HCM) then
            canbot = log(zkhcm/Z0GHCM)/log(1.0_rk/Z0GHCM)
        else if (ZK >= 0 .and. ZK <= z0g) then
            canbot = 0.0  !No-slip condition at surface (u=0 at z=0)
        else
            canbot = 1  ! should be zk > hcm
        end if
...
...
...
       if (ZK <= HCM) then       !at or below canopy top --> modify uc winds
            CANWIND = uc*canbot*cantop
        else
            CANWIND = UBZREF       !above canopy top constant and equal to reference winds
        end if

zmoon replied:
Well I didn't try it, but I suspect it would work automatically for profile (and still for point) if you made it an `elemental subroutine`. But if you would prefer just putting loop inside the subroutine, that's fine, I do think that is preferable to doing it outside.

drnimbusrain replied:
@zmoon OK.  I tried it and this seems like we should make this an idea for its own issue outside of this PR.  For canopy_wind,  I would have to change more of the code when making it an elemental subroutine and passing in whole array, because of the following errors:

canopy_wind_mod.F90:90:54:

   90 |             write(*,*) "critical problem: hcan <= zpd"
      |                                                      1
Error: IO UNIT in WRITE statement at (1) must be an internal file in a PURE procedure
canopy_wind_mod.F90:137:97:

  137 |                 write(*,*) 'wrong namelist option', RSL_OPT, 'only 0 (MOST) or 1 (RSL) available'
      |                                                                                                 1
Error: IO UNIT in WRITE statement at (1) must be an internal file in a PURE procedure
canopy_wind_mod.F90:149:86:

  149 |             write(*,*)  'Uc cannot be <= 0 at  ',uc , ' in canopy_wind calc...exiting'
      |                                                                                      1
Error: IO UNIT in WRITE statement at (1) must be an internal file in a PURE procedure
canopy_wind_mod.F90:91:24:

   91 |             call exit(2)
      |                        1
Error: Subroutine call to intrinsic ‘exit’ at (1) is not PURE
canopy_wind_mod.F90:138:28:

  138 |                 call exit(2)
      |                            1
Error: Subroutine call to intrinsic ‘exit’ at (1) is not PURE

zmoon replied:
Ah yes, `elemental` implies `pure` and printing not allowed in `pure`. So I guess putting the loop inside is the way to go in order to keep those messages. Separate PR fine for sure.

zmoon commented on README.md:
```diff
@@ -10,35 +10,97 @@ Compile, edit namelist, and run canopy model:
+    | ifcanphot        | logical canopy photolysis option (default = .FALSE.)                               |
+    | href             | real value of reference height above canopy associated with input wind speed (m)   |
+    | z0ghcm           | ratio of ground roughness length to canopy top height (Massman et al., 2017)       |
+    | lamdars          | Influence function associated with roughness sublayer (Massman et al., 2017)       |
```
"influence function"? seems like it's just a single number

zmoon replied:
(apologies if I commented on this elsewhere)

drnimbusrain replied:
Thanks Zach.   Yes, this is not a good description.  I have updated it in the next [commit](https://github.com/noaa-oar-arl/canopy-app/pull/20/commits/1d7e925674e0c389b6f21d55eaeb6b468e49796f).

zmoon replied:
:+1:

drnimbusrain commented on README.md:
```diff
@@ -10,35 +10,97 @@ Compile, edit namelist, and run canopy model:
+    | Variable Name    | Variable Descrip. (Units)                         |
+    | ---------------  | ------------------------------------------------- |
+    | CANLAY           | Canopy layer number (integer)                     |
+    | ZK               | Canopy layer resolution (m)                       |
```
Agree with both comments.  Deciding to remove the input_profile.txt file as it is not needed with canres and canlays as namelist inputs.

zmoon replied:
👍

drnimbusrain commented on canopy_zpd_mod.F90:
```diff
@@ -0,0 +1,84 @@
+        REAL(RK),    INTENT( IN )  :: ZTOTHC  ( : )   ! SUB-CANOPY Total z/h layers (nondimensional)
+        REAL(RK),    INTENT( IN )  :: FAFRACK ( : )   ! SUB-CANOPY Fractional (z) shapes of the
+        ! plant surface distribution at z=0 (nondimensional)
+        REAL(RK),    INTENT( IN )  :: FAFRACK0        ! Fractional (z) shapes of the
```
Agreed.  There is a typo in the first argument for FAFRACK (:), where I have removed the "at z=0 (nondimensional)".  Agreed also that FAFRACK0 is technically zero since there is no cumulative LAI here.  This corresponds to Eq. 12 and 15 in Massman et al., 2017.   

Also agree to remove FAFRACK0 since its not needed as the entire FAFRACK array is being passed in.  Also changed FAFRACK to FCLAI in this module per previous comments.  Changes are found in next [commit](https://github.com/noaa-oar-arl/canopy-app/pull/20/commits/048759144984452b20ada04c065064e37bad149f) if you agree to resolve conversation.

zmoon replied:
👍

drnimbusrain commented on canopy_zpd_mod.F90:
```diff
@@ -0,0 +1,84 @@
+!     IN/OUT
+        REAL(RK),    INTENT( IN )  :: ZTOTHC  ( : )   ! SUB-CANOPY Total z/h layers (nondimensional)
+        REAL(RK),    INTENT( IN )  :: FAFRACK ( : )   ! SUB-CANOPY Fractional (z) shapes of the
+        ! plant surface distribution at z=0 (nondimensional)
```
Agreed. There is a typo in the first argument for FAFRACK (:), where I have removed the "at z=0 (nondimensional)".  Changes are found in next [commit](https://github.com/noaa-oar-arl/canopy-app/pull/20/commits/048759144984452b20ada04c065064e37bad149f) if you agree to resolve conversation.

drnimbusrain commented on canopy_phot_mod.F90:
```diff
@@ -0,0 +1,53 @@
+        REAL(RK),    INTENT( IN )  :: LAI             ! Model input total Leaf Area Index
+        REAL(RK),    INTENT( IN )  :: CLU             ! Model input Clumping Index
+        REAL(RK),    INTENT( IN )  :: COSZEN          ! Model input Cosine Solar Zenith Angle
+        REAL(RK),    INTENT( OUT ) :: RJCORR          ! Photolysis correction factor
```
@zmoon Agreed, and changed in last [commit ](https://github.com/noaa-oar-arl/canopy-app/pull/20/commits/580eca3f68f58426c9ba66dc36a0fa8a66c4bf5b)given previous comments.

zmoon replied:
thanks

drnimbusrain commented on namelist.canopy:
```diff
@@ -4,17 +4,25 @@
+  ifcaneddy  = .TRUE.
+  ifcanphot  = .TRUE.
+  href       =  10.0
+  z0ghcm     =  0.0025
```
the "m" is for meters for dividing by the height of canopy (hc) in meters (m), i.e., hcm.

zmoon replied:
Ah I see, but since it is ratio of lengths, maybe the `m` not needed in the name?

drnimbusrain replied:
@zmoon Yes, true, I have changed it to z0ghc in the next [commit](https://github.com/noaa-oar-arl/canopy-app/pull/20/commits/3caa4dbe1a4fa528f9d1070a902b0764edadf9da).  Please resolve if OK.  Thanks!

zmoon replied:
👍

drnimbusrain commented on canopy_readnml.F90:
```diff
@@ -95,16 +101,31 @@ SUBROUTINE canopy_readnml (nlat,nlon,canlays,canres,href,z0ghcm,lamdars, &
     lamdars = 1.25
```
@zmoon Nice catch.  I have add rk for those with nonzero decimal portion in recent [commit](https://github.com/noaa-oar-arl/canopy-app/pull/20/commits/b955eaf8b2bb0f44a329c3e0294a57df3abb23ae).

zmoon replied:
👍

drnimbusrain commented on canopy_eddyx_mod.F90:
```diff
@@ -0,0 +1,111 @@
+contains
+
+!:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
+    SUBROUTINE CANOPY_EDDYX( HCM, ZK, USTAR, MOL, KZ )
```
@zmoon As in my previous response to inputting whole profile for the canopy_wind module, it seems this one needs specific HCM and ZK inputs in a loop somewhere because of the conditions in the code, for example:

       hol = HCM/MOL
        tlc = (HCM/USTAR) * (                        &
            (0.256 * (ZK-(0.75*HCM))/HCM ) +      &
            (0.492*EXP((-0.256*ZK/HCM)/0.492)) )
        sigma = 0.0_rk

        IF ( hol .LT. -0.1 )  THEN  !UNSTABLE
            IF ( ZK/HCM .GT. 1.25 ) THEN !SIGMACAN = Eulerian vertical velocity variance
                sigma = 1.25*USTAR
            END IF
            IF ( ZK/HCM .GE. 0.175  .AND.  ZK/HCM .LE. 1.25 ) THEN
                sigma = USTAR * ( 0.75 + (0.5 * COS((pi/1.06818) *     &
                    (1.25 - (ZK/HCM)))) )
            END IF
            IF ( ZK/HCM .LT. 0.175 )  THEN
                sigma = 0.25*USTAR
            END IF
        END IF

Unless you have a better idea, I leave this in the same format.

drnimbusrain replied:
@zmoon See my latest comment on using elemental subroutine in canopy_wind and the errors that result.  I think it may work better here, but if you agree, lets make a new issue to explore making all modules elemental to allow passing whole profiles when we improve codes after this PR.

zmoon replied:
Sounds fine. We can also fake elemental by putting the loop version in another subroutine and joining the two with an interface block, allowing it to be called both ways.

drnimbusrain commented on README.md:
```diff
@@ -10,35 +10,97 @@ Compile, edit namelist, and run canopy model:
+
+    | Variable Name    | Variable Descrip. (Units)                         |
+    | ---------------  | ------------------------------------------------- |
+    | CANLAY           | Canopy layer number (integer)                     |
```
@zmoon I have decided to remove the input profile file, and canlays is a user input to simply derive zk in the model. This was from your previous comment and associated [commit ](https://github.com/noaa-oar-arl/canopy-app/pull/20/commits/7c59fdd1976ab62f34f11a54028c953a779ac936).  Please let me know if we can resolve this, or if you think this one needs more work.  Seems that a simple user control of the canopy profile and derived zk (assumed constant dz at this point) is the better procedure to me, but please let me know if you disagree.

zmoon replied:
I think the way you went seems good for now 👍

drnimbusrain commented on canopy_phot_mod.F90:
```diff
@@ -0,0 +1,53 @@
+
+! Citation:
+!Makar, P., Staebler, R., Akingunola, A. et al. The effects of forest canopy shading and turbulence on boundary layer ozone.
+!Nat Commun 8, 15243 (2017). https://doi.org/10.1038/ncomms15243
```
I think I have addressed this in previous [commit](https://github.com/noaa-oar-arl/canopy-app/pull/20/commits/580eca3f68f58426c9ba66dc36a0fa8a66c4bf5b).  If so, we can resolve this conversation.

drnimbusrain commented on canopy_parm_mod.F90:
```diff
@@ -111,23 +110,26 @@ SUBROUTINE CANOPY_PARM( VTYPE, FCH, FFRAC, LAI, &
-                    PAI=LAI
-                else if (PAI_OPT .eq. 3) then !use set PAI value from user
+                else if (PAI_OPT .eq. 2) then !PAI = LAI + SAI (WAI)
+                    PAI=LAI + 0.52_rk  !WAI  = 0.52 from Toda and Richardson (2018):  https://doi.org/10.1016/j.agrformet.2017.09.004
```
@zmoon Yes, this is a big simplification.  There is not much data on PAI/WAI, so this is based on a WAI measurement during winter and leaf down period.  I add the Section number where I got this number and added new [commit](https://github.com/noaa-oar-arl/canopy-app/pull/20/commits/ded3d055b54461b013aed993177f153a512512ff). Please let me know if we can resolve this conversation.

zmoon replied:
👍

zmoon commented on canopy_phot_mod.F90:
```diff
@@ -0,0 +1,53 @@
+!     IN/OUT
+        REAL(RK),    INTENT( IN )  :: HCM             ! Height of canopy top (m)
+        REAL(RK),    INTENT( IN )  :: ZK              ! Above/Below canopy height, z (m)
+        REAL(RK),    INTENT( IN )  :: FAFRACK         ! Model input Fractional (z) shapes of the
```
sounds fine!

zmoon commented on README.md:
```diff
@@ -10,35 +10,87 @@ Compile, edit namelist, and run canopy model:
+    | ---------------  | ---------------------------------------------------------------------------------- |
+    | nlat             | number of latitude cells (must match # of LAT in `file_vars` above )               |
+    | nlon             | number of longitude cells (must match # of LON in `file_vars`above )               |
+    | canlays          | number of canopy layers                                                            |
```
doesn't the vertical grid include layers above the canopy as well? maybe `canlays` isn't the best name in that case

drnimbusrain replied:
I completely agree with you, and have changed the variable name to "modlays" if you agree in the next [commit](https://github.com/noaa-oar-arl/canopy-app/pull/20/commits/508b5f2f5008f59a7a2ef5298ee68c79e7c5a4ff).

drnimbusrain replied:
@zmoon Is this OK to resolve?  Note.  I have also changed "canres" to "modres".

drnimbusrain commented on README.md:
```diff
@@ -10,35 +10,97 @@ Compile, edit namelist, and run canopy model:
+    | CSZ              | Cosine of the solar zenith angle (dimensionless)  |
+    | Z0               | Total surface roughness length (m)                |
+    | MOL              | Monin-Obukhov Length (m)                          |
+    | FRP              | Total Fire Radiative Power (MW/grid cell)         |
```
@zmoon Yes, per grid cell area, fixed in next [commit](https://github.com/noaa-oar-arl/canopy-app/pull/20/commits/ab7601eea04f58dd15574187bbab55827bef13c2).  Please resolve if OK.

zmoon replied:
so it's not MW/m2 though? something like MW/m2 * grid cell area?

drnimbusrain commented on canopy_driver.F90:
```diff
@@ -113,11 +132,22 @@ program canopy_driver
+    if (ifcanphot) then
+        write(*,*)  'Canopy photolysis attenuation option selected'
+    end if
+    if (.not. ifcanwind .and. .not. ifcaneddy .and. .not. ifcanphot) then
```
@zmoon Yes, much better and changed in [commit](https://github.com/noaa-oar-arl/canopy-app/pull/20/commits/c4edffbc62950d8885d3f0192a277f77f8696201).  Please resolve if OK.

drnimbusrain commented on canopy_driver.F90:
```diff
@@ -113,11 +132,22 @@ program canopy_driver
+    end if
+    if (.not. ifcanwind .and. .not. ifcaneddy .and. .not. ifcanphot) then
+        write(*,*)  'No option(s) selected...see namelist'
+        call exit(2)
```
@zmoon Yes, much better and changed in [commit](https://github.com/noaa-oar-arl/canopy-app/pull/20/commits/c4edffbc62950d8885d3f0192a277f77f8696201). Please resolve if OK.

zmoon replied:
👍

drnimbusrain commented on canopy_driver.F90:
```diff
@@ -169,54 +222,112 @@ program canopy_driver
 
-! ... call canopy parameters to get canopy, fire info, and shape distribution parameters
+! ... get scaled canopy model profile and layers
+        ztothc      = zkcm/hcm
```
@zmoon Agreed.  Changed ztothc to zhc in [commit](https://github.com/noaa-oar-arl/canopy-app/pull/20/commits/3efad1a735f53d515d598d334dcdcc5fcc060624).  OK to resolve.

zmoon replied:
👍

drnimbusrain commented on canopy_driver.F90:
```diff
@@ -169,54 +222,112 @@ program canopy_driver
+
+! ... check for contiguous canopy conditions at each model grid cell
+            if (hcm .gt. fch_thresh .and. ffracref .gt. frt_thresh &
+                .and. lairef .ge. lai_thresh) then
```
@zmoon OK, this was legacy from my previous CMAQ implementation.  Agreed, its pretty arbitrary anyway.  I have changed to .gt. in [commit](https://github.com/noaa-oar-arl/canopy-app/pull/20/commits/2b90ef855f3faeeb76d3e0d071771ceba9f8f715).  OK to resolve.

drnimbusrain commented on canopy_driver.F90:
```diff
@@ -169,54 +222,112 @@ program canopy_driver
+                        write(*,*)  'Wrong FLAMEH_OPT choice of ', flameh_opt, ' in namelist...exiting'
+                        call exit(2)
+                    end if
+                    if (flameh .lt. canres) then !flameh under first layer height
```
@zmoon Yes, agreed this is a confusing description, and I have updated in next [commit](https://github.com/noaa-oar-arl/canopy-app/pull/20/commits/7539d8e04ecc031356e4d59503b35087a7f18087).   I leave it up to you if you agree my changes are clearer to resolve conversation.

zmoon replied:
👍

drnimbusrain commented on canopy_driver.F90:
```diff
@@ -169,54 +222,112 @@ program canopy_driver
+                    if (flameh .gt. 0.0) then !only calculate when flameh > 0
+                        call canopy_waf(hcm, lamdars, href, flameh, firetype, d_h, zo_h, &
+                            canBOT(midflamepoint), canTOP(midflamepoint), waf(loc))
+                    end if
```
@zmoon Well, in this case I don't think we need a warning, and definitely not an exit.  There are cases that canopy wind is calculated (under contiguous canopy conditions), but WAF is not (no fires FRP = 0).  We could print a warning for each grid cell without fire, but I don't think it is necessary as it results in long prints..."Warning...no fire and not calculating WAF."  Therefore, I will leave this as is.  Please resolve if you agree.

zmoon replied:
The condition is about the value of `flameh`, not about FRP. Does `flameh` being < 0 not indicate an issue?

zmoon replied:
I guess `flameh` will be 0 if `frp` is 0. I would make the condition about FRP then.

zmoon replied:
@drnimbusrain what do you think about that?

drnimbusrain replied:
@zmoon Actually though, there is a condition where flameh > 0 and there is no FRP.  This is the case when not running the WAF model with FRP at all, but rather if the user selected to input their own flameh in the namelist.  Maybe we should leave it as flameh condition.

drnimbusrain commented on canopy_eddyx_mod.F90:
```diff
@@ -0,0 +1,111 @@
+!-----------------------------------------------------------------------
+!-----------------------------------------------------------------------
+        use canopy_const_mod, ONLY: rk, pi !constants for canopy models
+        use canopy_utils_mod               !utilities for canopy models
```
Correct.  I have removed it in the next [commit](https://github.com/noaa-oar-arl/canopy-app/pull/20/commits/4f587aa9258fe9bb4a8ac3ceee7ef7ac1fe8a509).

drnimbusrain commented on canopy_eddyx_mod.F90:
```diff
@@ -0,0 +1,111 @@
+
+        hol = HCM/MOL
+        tlc = (HCM/USTAR) * (                        &
+            (0.256 * (ZK-(0.75*HCM))/HCM ) +      &
```
OK, added _rk to the float literals in next [commit](https://github.com/noaa-oar-arl/canopy-app/pull/20/commits/ea5625c4c3b17eb83149709a5a29c43db1d19d60).

drnimbusrain commented on canopy_eddyx_mod.F90:
```diff
@@ -0,0 +1,111 @@
+!Nat Commun 8, 15243 (2017). https://doi.org/10.1038/ncomms15243
+!Raupauch M. R. A Practical Lagrangian method for relating scalar concentrations to
+! source distributions in vegetation canopies. Q. J. R. Meteor. Soc. (1989), 115, pp 609-632
+
```
OK, added equation/page numbers for reference in next [commit](https://github.com/noaa-oar-arl/canopy-app/pull/20/commits/42c1eee018650590a516f2f0242c25a9dc9998d2).

drnimbusrain commented on canopy_utils_mod.F90:
```diff
@@ -70,5 +70,23 @@ function CalcPAI(fch, ffrac)
+        real(rk), intent(in)  :: dx           !! Input Grid cell length (m)
+        real(rk)              :: CalcFlameH   !! Calculated Plant area index (PAI)
+
+        CalcFlameH=0.0775*((frp*1000.0)/dx)**0.46  !Byram flameh calculation as function of FRP
```
OK, added _rk(s) to float literals in next [commit.](https://github.com/noaa-oar-arl/canopy-app/pull/20/commits/ffbe4427a5e62e113678eedabc00e388ed1e3ca7)

drnimbusrain commented on canopy_wind_mod.F90:
```diff
@@ -23,36 +24,49 @@ SUBROUTINE CANOPY_WIND( HCM, ZK, FAFRACK, UBZREF, Z0GHCM, &
 ! (2017)  W.J. Massman, J.M. Forthofer, M.A. Finney.  https://doi.org/10.1139/cjfr-2016-0354
         zkhcm = ZK/HCM
         z0g = Z0GHCM*HCM
+        hol = HCM/MOL
```
Correct, and I have removed the hol variable in next [commit](https://github.com/noaa-oar-arl/canopy-app/pull/20/commits/8c2a4628486ba17b29501b0b59a4f69ca2af2b57).

drnimbusrain commented on canopy_wind_mod.F90:
```diff
@@ -68,16 +82,84 @@ SUBROUTINE CANOPY_WIND( HCM, ZK, FAFRACK, UBZREF, Z0GHCM, &
+        z0m = ZO_H*HCM !aerodynamic roughness lengh (not scaled to HCM)
+        if((HCM-zpd) <= 0.) then
+            write(*,*) "critical problem: hcan <= zpd"
+            call exit(2)
```
OK, agree and made change in next [commit](https://github.com/noaa-oar-arl/canopy-app/pull/20/commits/1f0ff0728dd29779947c58b1abdea1cf4815f32d).

drnimbusrain commented on canopy_wind_mod.F90:
```diff
@@ -68,16 +82,84 @@ SUBROUTINE CANOPY_WIND( HCM, ZK, FAFRACK, UBZREF, Z0GHCM, &
+
+        if (uc <= 0.0) then
+            write(*,*)  'Uc cannot be <= 0 at  ',uc , ' in canopy_wind calc...exiting'
+            call exit(2)
```
OK, agree and made change in next [commit](https://github.com/noaa-oar-arl/canopy-app/pull/20/commits/9fc68de4216219e71cdb25d283d5882167e7f548).

drnimbusrain commented:
@zmoon Thank you for your excellent review!  I think I have addressed all your comments and made the necessary changes, with exception of implementing all submodules as elemental subroutines.  If you agree, we can resolve all conversations and merge this PR into Main, and add the elemental possibility as a new larger issue.  This way we can start to work on the issues to further improve these codes incrementally with your help.

zmoon commented on README.md:
```diff
@@ -57,11 +57,9 @@ Current Canopy-App inputs:
-
-
-    | Variable Name    | Variable Descrip. (Units)                                                          |
+    | Namelist Opt.    | Namelist Descript. and Units                                                       |
```
Was the change to "Descript." intended? The other one has "Descrip." still. There's probably enough room to spell out "Description".

drnimbusrain replied:
@zmoon , Sorry, let me fix this in the next [commit.](https://github.com/noaa-oar-arl/canopy-app/pull/20/commits/e331d0e9f1a79bef12127b7384f6ad0e5d5c88a9)

drnimbusrain commented:
Hey Zach

Yes,  this was a quick calc of DX, I'll add the haversine formula,  and
break this out into its own submodule

On Tue, Sep 27, 2022, 2:48 PM Zachary Moon ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In canopy_driver.F90
> <https://github.com/noaa-oar-arl/canopy-app/pull/20#discussion_r981600810>
> :
>
> > @@ -210,13 +211,32 @@ program canopy_driver
>                      z0ghcm, cdrag, pai, canBOT(i), canTOP(i), canWIND(i, loc))
>              end do
>
> -! ... calculate wind adjustment factor dependent on canopy winds and fire type
> -
> +! ... calculate wind adjustment factor dependent on fires (FRP), canopy winds,
> +!                                                        flameh, and fire type
> +            if (flameh_opt .eq. 0) then
> +                if (nlon .eq. 1) then      !single lon point -- reset to user value
> +                    flameh = flameh_set
> +                else                       !calculate flameh
> +                    dx     = dlon*111000.0  !convert lon grid to distance (m)
>
> Yes, it is more robust now. What I was alluding to though is that the
> value of dx in such a lat/lon grid changes with latitude. And we could
> account for that. But maybe that can wait until we set up to work with 2-d
> array inputs (e.g. loaded from netCDF files).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/noaa-oar-arl/canopy-app/pull/20#discussion_r981600810>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGLFYNSHLFQL4DC7QWKFE7LWAM6PLANCNFSM57MJFB4Q>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
